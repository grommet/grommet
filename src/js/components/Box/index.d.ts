import * as React from 'react';
import {
  A11yTitleType,
  AlignType,
  AlignContentType,
  AlignSelfType,
  BackgroundType,
  BasisType,
  BorderType,
  DirectionType,
  ElevationType,
  FillType,
  GapType,
  GridAreaType,
  HeightType,
  MarginType,
  PadType,
  PolymorphicType,
  RoundType,
  WidthType,
  AnimationType,
  SkeletonType,
} from '../../utils';

export interface BoxProps {
  a11yTitle?: A11yTitleType;
  alignSelf?: AlignSelfType;
  gridArea?: GridAreaType;
  margin?: MarginType;
  align?: AlignType;
  alignContent?: AlignContentType;
  animation?: AnimationType;
  background?: BackgroundType;
  basis?: BasisType;
  border?: BorderType;
  cssGap?: boolean;
  direction?: DirectionType;
  elevation?: ElevationType;
  flex?: 'grow' | 'shrink' | boolean | { grow?: number; shrink?: number };
  fill?: FillType;
  focusIndicator?: boolean;
  gap?: GapType | { row?: GapType; column?: GapType };
  height?: HeightType;
  hoverIndicator?:
    | { background?: BackgroundType; elevation?: ElevationType }
    | BackgroundType
    | boolean;
  justify?:
    | 'around'
    | 'between'
    | 'center'
    | 'end'
    | 'evenly'
    | 'start'
    | 'stretch';
  overflow?:
    | 'auto'
    | 'hidden'
    | 'scroll'
    | 'visible'
    | {
        horizontal?: 'auto' | 'hidden' | 'scroll' | 'visible';
        vertical?: 'auto' | 'hidden' | 'scroll' | 'visible';
      }
    | string;
  pad?: PadType;
  responsive?: boolean | 'container';
  round?: RoundType;
  skeleton?: SkeletonType;
  tag?: PolymorphicType;
  as?: PolymorphicType;
  width?: WidthType;
  wrap?: boolean | 'reverse';
}

export interface BoxExtendedProps
  extends BoxProps,
    Omit<JSX.IntrinsicElements['div'], keyof BoxProps> {}

// Keep type alias for backwards compatibility.
export type BoxTypes = BoxProps & JSX.IntrinsicElements['div'];

declare const Box: React.FC<BoxExtendedProps>;

export { Box };
