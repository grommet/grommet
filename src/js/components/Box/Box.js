import React, {
  Children,
  forwardRef,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { ThemeContext } from 'styled-components';
import {
  backgroundIsDark,
  useForwardedRef,
  deviceResponsive,
  getBreakpoint,
} from '../../utils';
import { Keyboard } from '../Keyboard';

import { StyledBox, StyledBoxGap } from './StyledBox';
import { BoxPropTypes } from './propTypes';
import { SkeletonContext, useSkeleton } from '../Skeleton';
import { AnnounceContext } from '../../contexts/AnnounceContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { ResponsiveContainerContext, ResponsiveContext } from '../../contexts';
import { useThemeValue } from '../../utils/useThemeValue';
import { supportsContainerQueries } from '../../utils/responsive';

// The value of ResponsiveContainerContext
// when we're within a responsive container
const responsiveContainerValue = true;

const Box = forwardRef(
  (
    {
      a11yTitle,
      background: backgroundProp,
      border,
      children,
      cssGap,
      direction = 'column',
      elevation, // munged to avoid styled-components putting it in the DOM
      fill, // munged to avoid styled-components putting it in the DOM
      gap,
      kind, // munged to avoid styled-components putting it in the DOM
      onBlur,
      onClick,
      onFocus,
      overflow, // munged to avoid styled-components putting it in the DOM
      responsive: responsiveProp = true,
      tag,
      as,
      wrap, // munged to avoid styled-components putting it in the DOM,
      width, // munged to avoid styled-components putting it in the DOM
      height, // munged to avoid styled-components putting it in the DOM
      tabIndex,
      skeleton: skeletonProp,
      ...rest
    },
    ref,
  ) => {
    const { theme, passThemeFlag } = useThemeValue();
    // boxOptions was created to preserve backwards compatibility but
    // should not be supported in v3
    const { box: boxOptions } = useContext(OptionsContext);

    console.log('as', as, typeof as);
    const skeleton = useSkeleton();

    const responsiveContainer = useContext(ResponsiveContainerContext);
    const responsive =
      responsiveContainer && responsiveProp ? 'container' : responsiveProp;

    let background = backgroundProp;

    const announce = useContext(AnnounceContext);

    const [containerBreakpoint, setContainerBreakpoint] = useState(
      deviceResponsive(navigator.userAgent, theme) ||
        theme?.global?.deviceBreakpoints.tablet,
    );

    const containerRef = useForwardedRef(ref);
    // console.log('render', responsiveProp, containerRef);

    // Update the size in the ResponsiveContext at this
    // level if we're in responsive container mode.
    useEffect(() => {
      // console.log('usEffect', responsiveProp, containerRef);
      const containerElement = containerRef?.current;
      if (responsiveProp !== 'container' || !containerElement) return undefined;

      let resizeObserver;
      if (typeof window !== 'undefined' && window.ResizeObserver) {
        resizeObserver = new window.ResizeObserver((entries) => {
          window.requestAnimationFrame(() => {
            if (!Array.isArray(entries) || !entries.length) {
              return;
            }
            const size = entries[0]?.borderBoxSize?.[0]?.inlineSize;
            // console.log('Observer size', size, getBreakpoint(size, theme));
            if (size) {
              setContainerBreakpoint(getBreakpoint(size, theme));
            }
          });
        });

        resizeObserver.observe(containerElement);
      }
      // Initial size for ResponsiveContext in responsive container mode.
      // Also is a fallback for server side rendering
      const containerWidth = containerElement.getBoundingClientRect().width;
      setContainerBreakpoint(getBreakpoint(containerWidth, theme));

      return () => {
        if (resizeObserver) {
          resizeObserver.disconnect();
        }
      };
    }, [theme, responsiveProp, containerRef]);

    useEffect(() => {
      if (skeletonProp?.message?.start) announce(skeletonProp.message.start);
      else if (typeof skeletonProp?.message === 'string')
        announce(skeletonProp.message);
      return () =>
        skeletonProp?.message?.end && announce(skeletonProp.message.end);
    }, [announce, skeletonProp]);

    const focusable = useMemo(
      () => onClick && !(tabIndex < 0),
      [onClick, tabIndex],
    );

    const [focus, setFocus] = useState();

    const clickProps = useMemo(() => {
      if (focusable) {
        return {
          onClick,
          onFocus: (event) => {
            setFocus(true);
            if (onFocus) onFocus(event);
          },
          onBlur: (event) => {
            setFocus(false);
            if (onBlur) onBlur(event);
          },
        };
      }
      const result = {};
      if (onBlur) result.onBlur = onBlur;
      if (onClick) result.onClick = onClick;
      if (onFocus) result.onFocus = onFocus;
      return result;
    }, [focusable, onClick, onFocus, onBlur]);

    const adjustedTabIndex = useMemo(() => {
      if (tabIndex !== undefined) return tabIndex;
      if (focusable) return 0;
      return undefined;
    }, [focusable, tabIndex]);

    if (
      (border === 'between' ||
        (border && border.side === 'between') ||
        (Array.isArray(border) && border.find((b) => b.side === 'between'))) &&
      !gap
    ) {
      console.warn('Box must have a gap to use border between');
    }

    let contents = children;
    if (
      gap &&
      gap !== 'none' &&
      (!(boxOptions?.cssGap || cssGap || typeof gap === 'object') ||
        // need this approach to show border between
        border === 'between' ||
        border?.side === 'between' ||
        (Array.isArray(border) && border.find((b) => b.side === 'between')))
    ) {
      // if border is an array, we need to extract the border between object
      const styledBoxGapBorder = Array.isArray(border)
        ? border.find((b) => b.side === 'between')
        : border;
      const boxAs = !as && tag ? tag : as;
      contents = [];
      let firstIndex;
      Children.forEach(children, (child, index) => {
        if (child) {
          if (firstIndex === undefined) {
            firstIndex = index;
          } else {
            contents.push(
              <StyledBoxGap
                // eslint-disable-next-line react/no-array-index-key
                key={`gap-${index}`}
                as={boxAs === 'span' ? boxAs : 'div'}
                gap={gap}
                directionProp={direction}
                responsive={responsive}
                border={styledBoxGapBorder}
                {...passThemeFlag}
              />,
            );
          }
        }
        contents.push(child);
      });
    }

    const nextSkeleton = useMemo(() => {
      // Decide if we need to add a new SkeletonContext. We need one if:
      //   1. skeleton info was set in a property OR
      //   2. there already is a SkeletonContext but this box has a
      //      background or border. This means the box probably is more
      //      distinguishable from the area around it.
      // We keep track of a depth so we know how to alternate backgrounds.
      if (skeletonProp || ((background || border) && skeleton)) {
        const depth = skeleton ? skeleton.depth + 1 : 0;
        return {
          ...skeleton,
          depth,
          ...(typeof skeletonProp === 'object' ? skeletonProp : {}),
        };
      }
      return undefined;
    }, [background, border, skeleton, skeletonProp]);

    let skeletonProps = {};
    if (nextSkeleton) {
      const {
        colors: skeletonThemeColors,
        size: skeletonThemeSize,
        ...skeletonThemeProps
      } = theme.skeleton;
      const skeletonColors = nextSkeleton.colors
        ? nextSkeleton.colors[theme.dark ? 'dark' : 'light']
        : skeletonThemeColors?.[theme.dark ? 'dark' : 'light'];
      skeletonProps = { ...skeletonThemeProps };
      background = skeletonColors[nextSkeleton.depth % skeletonColors.length];
      if (skeletonProp?.animation) {
        skeletonProps.animation = skeletonProp.animation;
      }
      contents = (
        <SkeletonContext.Provider value={nextSkeleton}>
          {contents}
        </SkeletonContext.Provider>
      );
    }

    // construct a new theme object in case we have a background that wants
    // to change the background color context
    const nextTheme = useMemo(() => {
      let result;
      if (background || theme.darkChanged) {
        const dark = backgroundIsDark(background, theme);
        const darkChanged = dark !== undefined && dark !== theme.dark;
        if (darkChanged || theme.darkChanged) {
          result = { ...theme };
          result.dark = dark === undefined ? theme.dark : dark;
          result.background = background;
        } else if (background) {
          // This allows DataTable to intelligently set the background
          // of a pinned header or footer.
          result = { ...theme };
          result.background = background;
        }
      }
      return result || theme;
    }, [background, theme]);

    let content = (
      <StyledBox
        as={!as && tag ? tag : as}
        aria-label={a11yTitle}
        background={background}
        border={border}
        ref={containerRef}
        directionProp={direction}
        elevationProp={elevation}
        fillProp={fill}
        focus={focus}
        gap={
          (boxOptions?.cssGap || cssGap || typeof gap === 'object') &&
          gap &&
          gap !== 'none' &&
          border !== 'between' &&
          border?.side !== 'between' &&
          (!Array.isArray(border) ||
            !border.find((b) => b.side === 'between')) &&
          gap
        }
        kindProp={kind}
        overflowProp={overflow}
        wrapProp={wrap}
        widthProp={width}
        heightProp={height}
        responsive={responsive}
        responsiveContainer={responsiveProp === 'container'}
        tabIndex={adjustedTabIndex}
        {...clickProps}
        {...passThemeFlag}
        {...rest}
        {...skeletonProps}
      >
        <ThemeContext.Provider value={nextTheme}>
          {contents}
        </ThemeContext.Provider>
      </StyledBox>
    );

    if (responsiveProp === 'container') {
      if (supportsContainerQueries()) {
        content = (
          <ResponsiveContext.Provider value={containerBreakpoint}>
            <ResponsiveContainerContext.Provider
              value={responsiveContainerValue}
            >
              {content}
            </ResponsiveContainerContext.Provider>
          </ResponsiveContext.Provider>
        );
      } else {
        console.warn(
          '<Box responsive="container"> requires styled-components v6 or later',
        );
      }
    }

    if (onClick) {
      content = <Keyboard onEnter={onClick}>{content}</Keyboard>;
    }

    return content;
  },
);

Box.displayName = 'Box';
Box.propTypes = BoxPropTypes;
export { Box };
