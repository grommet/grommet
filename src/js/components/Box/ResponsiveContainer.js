import React, { useContext, useEffect, useRef, useState } from 'react';
import {
  ResponsiveContainerContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';
import { StyledResponsiveContainer } from './StyledResponsiveContainer';
import { deviceResponsive, getBreakpoint } from '../../utils';

const responsiveContainerValue = true;

export const ResponsiveContainer = ({ children }) => {
  const theme = useContext(ThemeContext);
  const ref = useRef(undefined);
  const [stateResponsive, setResponsive] = useState();

  useEffect(() => {
    const element = ref?.current;

    if (!element) return undefined;

    let resizeObserver;
    if (typeof window !== 'undefined' && window.ResizeObserver) {
      resizeObserver = new window.ResizeObserver((entries) => {
        window.requestAnimationFrame(() => {
          if (!Array.isArray(entries) || !entries.length) {
            return;
          }
          const size = entries[0]?.borderBoxSize?.[0]?.inlineSize;
          if (size) {
            setResponsive(getBreakpoint(size, theme));
          }
        });
      });

      resizeObserver.observe(element);
    } else {
      // fallback for server side rendering
      const { width } = element.getBoundingClientRect();
      setResponsive(getBreakpoint(width, theme));
    }
    return () => {
      if (resizeObserver) {
        resizeObserver.disconnect();
      }
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(navigator.userAgent, theme) ||
    theme?.global?.deviceBreakpoints.tablet;

  return (
    <ResponsiveContext.Provider value={responsive}>
      <ResponsiveContainerContext.Provider value={responsiveContainerValue}>
        <StyledResponsiveContainer ref={ref}>
          {children}
        </StyledResponsiveContainer>
      </ResponsiveContainerContext.Provider>
    </ResponsiveContext.Provider>
  );
};
