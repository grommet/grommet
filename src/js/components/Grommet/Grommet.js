import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  deviceResponsive,
  getBreakpoint,
  normalizeColor,
  useForwardedRef,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const defaultOptions = {};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    options = defaultOptions,
    messages: messagesProp,
    onAnalytics,
    ...rest
  } = props;
  const { background, dir, themeMode, userAgent } = props;
  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (
      themeProp &&
      themeProp.menu &&
      themeProp.menu.drop &&
      themeProp.menu.drop.align
    ) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';

    if (
      themeMode === 'auto' &&
      typeof window !== 'undefined' &&
      window.matchMedia &&
      window.matchMedia('(prefers-color-scheme: dark)').matches
    ) {
      nextTheme.dark = true;
    }

    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  const messages = useMemo(() => {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    const nextMessages = deepMerge(
      defaultMessages,
      messagesProp?.messages || {},
    );
    return {
      messages: nextMessages,
      format: (opts) => {
        const message = messagesProp?.format && messagesProp.format(opts);
        return typeof message !== 'undefined'
          ? message
          : format(opts, nextMessages);
      },
    };
  }, [messagesProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  const grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  const roots = useRef([]);
  useEffect(() => {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  const rootsContextValue = useMemo(() => ({ roots }), []);

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <RootsContext.Provider value={rootsContextValue}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <OptionsContext.Provider value={options}>
              <MessageContext.Provider value={messages}>
                <AnalyticsProvider onAnalytics={onAnalytics}>
                  <StyledGrommet full={full} {...rest} ref={grommetRef}>
                    {children}
                  </StyledGrommet>
                  {full && <FullGlobalStyle />}
                </AnalyticsProvider>
              </MessageContext.Provider>
            </OptionsContext.Provider>
          </ContainerTargetContext.Provider>
        </RootsContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;

export { Grommet };
