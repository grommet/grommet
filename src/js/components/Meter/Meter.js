import React, { forwardRef, useContext, useMemo } from 'react';

import { Bar } from './Bar';
import { Circle } from './Circle';
import { MeterPropTypes } from './propTypes';
import { useThemeValue } from '../../utils/useThemeValue';
import { MessageContext } from '../../contexts/MessageContext';

const Meter = forwardRef(
  (
    {
      'aria-label': ariaLabel,
      background = { color: 'light-2', opacity: 'medium' },
      color,
      direction = 'horizontal',
      max: maxProp,
      messages,
      size = 'medium',
      thickness = 'medium',
      type = 'bar',
      reverse: reverseProp,
      value,
      values: valuesProp,
      ...rest
    },
    ref,
  ) => {
    const { theme } = useThemeValue();
    const { format } = useContext(MessageContext);

    // normalize values to an array of objects
    const values = useMemo(() => {
      if (valuesProp) return valuesProp;
      if (value) return [{ color, value }];
      return [];
    }, [color, value, valuesProp]);

    const reverse =
      direction === 'horizontal' &&
      (theme.dir === 'rtl' || reverseProp) &&
      !(theme.dir === 'rtl' && reverseProp);

    const max = useMemo(() => {
      let maxValue = 100;
      if (values?.length > 1) {
        maxValue = values.reduce(
          (total, currentValue) => total + currentValue.value,
          0,
        );
      }
      return maxProp || maxValue || 100;
    }, [maxProp, values]);

    const messageId = values?.length === 1 ? 'singular' : 'plural';

    const meterType = type || 'bar';

    const meterAriaLabel =
      ariaLabel ||
      format({
        id: `meter.${meterType}.${messageId}`,
        messages: messages?.meter?.[meterType],
        values: {
          meterValue:
            value || values.map((item) => item.value ?? 0).join(', ') || 0,
          type,
          max,
        },
      });

    let content;
    if (type === 'bar') {
      content = (
        <Bar
          ref={ref}
          aria-label={meterAriaLabel}
          max={max}
          values={values}
          size={size}
          thickness={thickness}
          background={background}
          direction={direction}
          reverse={reverse}
          {...rest}
        />
      );
    } else if (type === 'circle' || type === 'pie' || type === 'semicircle') {
      content = (
        <Circle
          aria-label={meterAriaLabel}
          ref={ref}
          max={max}
          values={values}
          size={size}
          thickness={thickness}
          type={type}
          background={background}
          reverse={reverse}
          {...rest}
        />
      );
    }
    return content;
  },
);

Meter.displayName = 'Meter';
Meter.prototype = MeterPropTypes;

export { Meter };
