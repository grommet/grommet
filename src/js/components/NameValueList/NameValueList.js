import React, { forwardRef, useContext, useMemo } from 'react';
import { Grid } from '../Grid';
import { ResponsiveContext } from '../../contexts/ResponsiveContext';
import { NameValueListContext } from './NameValueListContext';
import { isSmall } from '../../utils/responsive';
import { useThemeValue } from '../../utils/useThemeValue';

const NameValueList = forwardRef(
  (
    {
      align,
      layout = 'column',
      nameProps,
      pairProps = { direction: 'row' },
      valueProps,
      ...rest
    },
    ref,
  ) => {
    const size = useContext(ResponsiveContext);
    const { theme } = useThemeValue();

    // If layout is grid, valueWidth sets the max width of the column.
    // Grid will 'fit' as many columns of valueWidth per row as container's
    // width allows.
    let columns;
    let valueWidth = valueProps?.width || theme.nameValueList.value.width;
    let nameWidth = nameProps?.width || theme.nameValueList.name.width;
    const formatWidth = (width) =>
      typeof width === 'object' ? [width.min, width.max] : width;

    nameWidth = formatWidth(nameWidth);
    valueWidth = formatWidth(valueWidth);

    if (isSmall(size) || layout === 'grid')
      columns = {
        count: 'fit',
        size: !Array.isArray(valueWidth) ? ['auto', valueWidth] : valueWidth,
      };
    else if (layout === 'column' && pairProps.direction === 'row')
      columns = [
        nameWidth,
        !Array.isArray(valueWidth) ? ['auto', valueWidth] : valueWidth,
      ];
    else columns = [valueWidth];

    let { gap } = theme.nameValueList;
    if (
      (pairProps.direction === 'column' || isSmall(size)) &&
      theme.nameValueList.pair?.column?.gap
    ) {
      gap = theme.nameValueList.pair.column.gap;
    }

    const listContextValue = useMemo(
      () => ({ nameProps, pairProps, valueProps }),
      [nameProps, pairProps, valueProps],
    );

    return (
      <NameValueListContext.Provider value={listContextValue}>
        <Grid
          as="dl"
          ref={ref}
          align={align}
          columns={columns}
          gap={gap}
          margin="none" // override browser default margin for dl
          {...rest}
        />
      </NameValueListContext.Provider>
    );
  },
);

NameValueList.displayName = 'NameValueList';

export { NameValueList };
