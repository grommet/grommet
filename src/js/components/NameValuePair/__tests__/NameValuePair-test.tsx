import React from 'react';
import 'jest-styled-components';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';
import { hpe } from 'grommet-theme-hpe';

import { render } from '@testing-library/react';

import { Box } from '../../Box';
import { Grommet } from '../../Grommet';
import { Meter } from '../../Meter';
import { NameValueList } from '../../NameValueList';
import { NameValuePair } from '..';

const data = {
  name: 'entry',
  location: 'San Francisco',
  health: 80,
};

describe('NameValuePair', () => {
  test(`should render name when name is typeof string`, () => {
    const { container } = render(
      <Grommet theme={hpe}>
        <NameValueList>
          {Object.entries(data).map(([name, value]) => (
            <NameValuePair key={name} name={name}>
              {value}
            </NameValuePair>
          ))}
        </NameValueList>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test(`should render name when name is JSX Element`, () => {
    const { container } = render(
      <Grommet>
        <NameValueList nameProps={{ width: 'xsmall' }}>
          {Object.entries(data).map(([name, value]) => (
            <NameValuePair
              key={name}
              name={
                <Box pad="small" background="brand">
                  {name}
                </Box>
              }
            >
              {value}
            </NameValuePair>
          ))}
        </NameValueList>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test(`should render value when provided as child of type 
  string or number`, () => {
    const { container } = render(
      <Grommet>
        <NameValueList valueProps={{ width: 'xsmall' }}>
          {Object.entries(data).map(([name, value]) => (
            <NameValuePair key={name} name={name}>
              {value}
            </NameValuePair>
          ))}
        </NameValueList>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test(`should render value when provided as child of type JSX Element`, () => {
    const { container } = render(
      <Grommet theme={hpe}>
        <NameValueList nameProps={{ align: 'end' }}>
          {Object.entries(data).map(([name, value]) => {
            let renderedValue: any = value;
            if (name === 'health' && typeof value === 'number') {
              renderedValue = <Meter value={value} />;
            }
            return (
              <NameValuePair key={name} name={name}>
                {renderedValue}
              </NameValuePair>
            );
          })}
        </NameValueList>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
