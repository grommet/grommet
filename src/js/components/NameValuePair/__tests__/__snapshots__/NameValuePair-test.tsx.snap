// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameValuePair should render name when name is JSX Element 1`] = `
.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #7D4CDB;
  color: #f8f8f8;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  padding: 12px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: grid;
  box-sizing: border-box;
  margin: 0px;
  grid-template-columns: repeat( auto-fit,minmax(auto, 384px) );
  grid-gap: 24px 48px;
}

.c4 {
  margin: 0px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
}

@media only screen and (max-width: 768px) {
  .c3 {
    padding: 6px;
  }
}

@media only screen and (max-width: 768px) {

}

@media only screen and (max-width: 768px) {
  .c1 {
    margin: 0px;
  }
}

<div
  class="c0"
>
  <dl
    class="c1"
  >
    <div
      class="c2"
    >
      <dt
        class="c2"
      >
        <div
          class="c3"
        >
          name
        </div>
      </dt>
      <dd
        class="c4"
      >
        entry
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c2"
      >
        <div
          class="c3"
        >
          location
        </div>
      </dt>
      <dd
        class="c4"
      >
        San Francisco
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c2"
      >
        <div
          class="c3"
        >
          health
        </div>
      </dt>
      <dd
        class="c4"
      >
        80
      </dd>
    </div>
  </dl>
</div>
`;

exports[`NameValuePair should render name when name is typeof string 1`] = `
.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: grid;
  box-sizing: border-box;
  margin: 0px;
  grid-template-columns: repeat( auto-fit,minmax(auto, 384px) );
  grid-gap: 24px 48px;
}

.c4 {
  margin: 0px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
}

.c3 {
  margin-bottom: 3px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
  font-weight: bold;
}

@media only screen and (max-width: 768px) {

}

@media only screen and (max-width: 768px) {
  .c1 {
    margin: 0px;
  }
}

<div
  class="c0"
>
  <dl
    class="c1"
  >
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        name
      </dt>
      <dd
        class="c4"
      >
        entry
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        location
      </dt>
      <dd
        class="c4"
      >
        San Francisco
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        health
      </dt>
      <dd
        class="c4"
      >
        80
      </dd>
    </div>
  </dl>
</div>
`;

exports[`NameValuePair should render value when provided as child of type 
  string or number 1`] = `
.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: grid;
  box-sizing: border-box;
  margin: 0px;
  grid-template-columns: repeat( auto-fit,minmax(auto, 96px) );
  grid-gap: 24px 48px;
}

.c4 {
  margin: 0px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
}

.c3 {
  margin-bottom: 3px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
  font-weight: bold;
}

@media only screen and (max-width: 768px) {

}

@media only screen and (max-width: 768px) {
  .c1 {
    margin: 0px;
  }
}

<div
  class="c0"
>
  <dl
    class="c1"
  >
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        name
      </dt>
      <dd
        class="c4"
      >
        entry
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        location
      </dt>
      <dd
        class="c4"
      >
        San Francisco
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        health
      </dt>
      <dd
        class="c4"
      >
        80
      </dd>
    </div>
  </dl>
</div>
`;

exports[`NameValuePair should render value when provided as child of type JSX Element 1`] = `
.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin: 0px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c6 {
  max-width: 100%;
  overflow: hidden;
}

.c6 path {
  transition: stroke 0.3s,stroke-width 0.3s;
}

.c1 {
  display: grid;
  box-sizing: border-box;
  margin: 0px;
  grid-template-columns: repeat( auto-fit,minmax(auto, 384px) );
  grid-gap: 24px 48px;
}

.c4 {
  margin: 0px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
}

.c3 {
  margin-bottom: 3px;
  font-size: 18px;
  line-height: 24px;
  color: #444444;
  font-weight: bold;
}

@media only screen and (max-width: 768px) {
  .c5 {
    margin: 0px;
  }
}

@media only screen and (max-width: 768px) {

}

@media only screen and (max-width: 768px) {
  .c1 {
    margin: 0px;
  }
}

<div
  class="c0"
>
  <dl
    class="c1"
  >
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        name
      </dt>
      <dd
        class="c4"
      >
        entry
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        location
      </dt>
      <dd
        class="c4"
      >
        San Francisco
      </dd>
    </div>
    <div
      class="c2"
    >
      <dt
        class="c3"
      >
        health
      </dt>
      <dd
        class="c5"
      >
        <svg
          aria-label="Bar meter with value 80 out of 100"
          class="c6"
          height="24"
          preserveAspectRatio="none"
          viewBox="0 0 384 24"
          width="384"
        >
          <path
            d="M 0,12 L 384,12"
            fill="none"
            stroke="#F2F2F2"
            stroke-linecap="square"
            stroke-opacity="0.4"
            stroke-width="24"
          />
          <path
            d="M 0,12 L 307.2,12"
            fill="none"
            stroke="#6FFFB0"
            stroke-linecap="butt"
            stroke-width="24"
          />
        </svg>
      </dd>
    </div>
  </dl>
</div>
`;
